trimming: Support finalizers

JeffBezanson opened the pull request:
This is sufficient to get finalizers working. Before merging this though I propose generalizing it by adding a new builtin `calledge` (or something) that we'd use like so:

```
function finalizer(@nospecialize(f), @nospecialize(o))
    _check_mutable(o)
    Core.finalizer(f, o)
    calledge(Tuple{typeof(f), typeof(o)})
    return o
end

function Core.Task(@nospecialize(f), reserved_stack::Int=0)
    calledge(Tuple{typeof(f)})
    Core._Task(f, reserved_stack, ThreadSynchronizer())
end
```

which says that if a call to `finalizer` is reachable, then so is `f(o)` etc. The implementation would be basically identical to what's here currently.

topolarity committed:
trim: Add `Core.finalizer` support

This adds the changes required to guess the MethodInstance that the
runtime will eventually invoke for this call, enqueue it for codegen,
and then verify its presence in the verifier.

JeffBezanson committed:
some fixes

topolarity committed:
typeinfer: add separate compile workqueue for native / call_latest code

topolarity commented on Compiler/src/typeinfer.jl:
```diff
@@ -1267,6 +1288,31 @@ function collectinvokes!(wq::Vector{CodeInstance}, ci::CodeInfo)
+                    mi === nothing && continue
+
+                    if mi.def.primary_world <= Base.get_world_counter() <= mi.def.deleted_world
+                        ci = typeinf_ext(interp, mi, SOURCE_MODE_GET_SOURCE)
+                        # TODO: separate workqueue for NativeInterpreter
+                        ci isa CodeInstance && push!(wq, ci)
```
This needs re-work. We need to at least check that we're not re-inserting an `mi` to the workqueue or else we can infinite loop because we'll get a fresh `ci` and continually re-insert a recursive finalizer

Jameson suggested a re-factor of the workqueue to accommodate this, which I've started on but it's not done just yet

topolarity commented:
> adding a new builtin calledge (or something) that we'd use like so

I was thinking perhaps `Core._predeclare_call(foo, a, b)`

topolarity committed:
Make `empty!(::BitSet)` work during bootstrap

topolarity committed:
Collect invoke only for `ftyp === typeof(finalizer)`

Otherwise this will match `ftyp === Builtin`, which is pretty silly
since that could just as well be any other built-in.

vchuravy commented on Compiler/src/typeinfer.jl:
```diff
@@ -1257,15 +1257,100 @@ function typeinf_type(interp::AbstractInterpreter, mi::MethodInstance)
+
+            ftyp = widenconst(argextype(farg, ci, sptypes))
+
+            if ftyp === typeof(Core.finalizer) && length(stmt.args) == 3
+                finalizer_fn = argextype(stmt.args[2], ci, sptypes)
+                object = argextype(stmt.args[3], ci, sptypes)
+                atype = argtypes_to_type(Any[finalizer_fn, object])
```
Could we instead use the existing `CallInfo` pieces? E.g. We already have a `FinalizerInfo` https://github.com/JuliaLang/julia/blob/b265dbac3ef8d1f8f7d779044f458a3c5e2b4689/Compiler/src/stmtinfo.jl#L455 .

I am thinking something like:
```
indirect(::CallInfo) = false
indirect(::FinalizerInfo) = true

if indirect(callinfo)
    enqueue(workqueue, callinfo.call)
end
```

and then we could add `TaskInfo` and so forth.

One note is https://github.com/JuliaLang/julia/blob/b265dbac3ef8d1f8f7d779044f458a3c5e2b4689/Compiler/src/stmtinfo.jl#L460 

Maybe we "just" need to change that to actually add the edge?

topolarity replied:
IIUC the optimizer destroys the `CallInfo` so we unfortunately can't forward this edge from inference in the usual way

vchuravy commented:
I am still not convinced of the utility of a new builtin like `predeclare_call`, that feels unnecessary since inference ought to have all the information already. 

If we truly need it, could we merge it into `precompile`?

topolarity commented:
> I am still not convinced of the utility of a new builtin like `predeclare_call`, that feels unnecessary since inference ought to have all the information already.

I'm not super pleased about it either - but I also don't know of an alternative for `atexit`-like patterns:
```julia
const atexit_handlers = Callable[]
push!(atexit_handlers, f) # we know the identity of `f` on the `push!`

run_atexit() = for handler in atexit_handlers
    handler() # we know longer know the identity of the handlers here
end
```

This is pretty challenging, and also very common in practice. Even if we add this "virtual edge" to the runtime (via `predeclare_call` or similar), it is not tracked properly if you happen to serialize an object that needs this edge tracked

(that happens to not be the case for `atexit` since those are cleared, but other type-erased callbacks like `LazyString` could easily be carried over from pre-compilation)

> If we truly need it, could we merge it into `precompile`?

Perhaps. That's what I called it in the first prototype, but @JeffBezanson has been keeping the `entrypoint` stuff separate for now, since there's little trade-off to declaring a `precompile` but potentially quite a lot of cost for declaring an `entrypoint` (or a "virtual edge") since it bloats `--trim` output and grows the body of static-dispatch-only code

vchuravy commented:
So the problem is two-fold?

1. The expression of a virtual edge for inference sake? 
2. Turning said virtual edge into an entry edge for trimming?

My expectations was that we could just turn these missing edges in inference into real edges there, e.g. teach it to look at the Task constructor and treat it as if it is a call, and that ought to add it to its real edge list? And that one could use the precompile built-in to express such a virtual edge.

For cases like atexit,  the problem is more that we have two "separate" inference entries, and their is indeed no information flow possible. I agree that in that case we need to mark these functions as entrypoints, but that seems independent for me from how we handle these calls when they are visible in inference?

vtjnash commented:
> My expectations was that we could just turn these missing edges in inference into real edges there, e.g. teach it to look at the Task constructor and treat it as if it is a call, and that ought to add it to its real edge list? And that one could use the precompile built-in to express such a virtual edge.

We could do something more like you describe, but it seemed likely to end up being more complicated than is useful. We don't actually want to compile most "real edges" here, we only want to compile "invoke edges" which are often derived from inlined or de-virtualized copies instead. Those edges also run under a different interpreter than the parent (NativeInterp @ current-world, never the inherited one), so it is inconvenient to have to pass around multiple interpreters inside of each NativeInterp in the off-chance that someone is running a trim build and will want to recover that information later.

Since this code is trim-relevant, we already know the code to precisely compute the call signature is already required (we need to check that every dispatch is covered or is a builtin function not on the disallow-list), so checking that this is one of a set of specific functions is already visible to the code here.

The tricky case is ones like `atexit`, where we do not easily have the tools to correctly express that it should expect the possibility of a runtime dispatch later for the lifetime of the object it was based upon (which might have come from either compile or runtime or even earlier in the sysimg creation). Inside a single process, this would be adequately modeled already by OpaqueClosure (fixed world hidden edge) or struct CFunction (dynamic world hidden edge), but neither of those survive serialization at present.

topolarity committed:
typeinfer: Work around inference bug

See https://github.com/JuliaLang/julia/issues/58143 for details.

topolarity commented:
@JeffBezanson this should be ready to go from my end

I'm happy to remove the `Core._predeclare_all` builtin for now so that we can tackle use case that separately - I'll defer to others to make that call.

AFAIU, that's only immediately useful for `Task` anyway, since we have no way to use this except for call-sites that Inference cannot see.

topolarity committed:
Adjust `--trim` tests for new functionality

JeffBezanson commented on base/idset.jl:
```diff
@@ -92,8 +92,17 @@ function sizehint!(s::IdSet, newsz)
+    t = @_gc_preserve_begin a
+    p = unsafe_convert(Ptr{Cvoid}, a)
+    T = eltype(a)
+    memset(p, 0x0, (sizeof(T) * length(a)) % UInt)
```
This seems like an odd place to put this optimization. Is it really that helpful? If so, should it be added to `fill!`? Or we could add a proper `zero!` function.

JeffBezanson replied:
Oh I see now this was for bootstrapping. Rest of the question still applies though,

topolarity replied:
The `fill!` implementation we need here is from `multidimensional.jl` since it's for a `Memory`:
```julia
julia> @which fill!(m, 0)
fill!(A::AbstractArray{T}, x) where T
     @ Base multidimensional.jl:1141
 ```

JeffBezanson replied:
Ah. Seems to me that should be in abstractarray.jl.

JeffBezanson commented:
If we're going to have predeclare_call (and I think we should) then we should use it to handle finalizer, Task, and atexit. Is there a compelling reason to make finalizers special but not other cases?

JeffBezanson commented:
OK I'm kind of sympathetic to the "polluted IR" argument. Having to manually add each case to the compiler when it could just be one line of code as needed seems so unfortunate though. Are we sure there won't be more cases?

I guess with `atexit` the problem is the call site for the exit handlers. There we need a different feature, which I think has been referred to as "unsafe_call", AKA a "trust me bro" call site. For that it seems having something in the IR is unavoidable?

vtjnash commented on Compiler/src/typeinfer.jl:
```diff
@@ -1533,9 +1599,44 @@ function typeinf_ext_toplevel(methods::Vector{Any}, worlds::Vector{UInt}, trim_m
-        latest = false
+        is_latest_world = false
     end
+
```
```suggestion
            compile!(codeinfos, invokelatest_queue; invokelatest_queue)
```

vtjnash commented on Compiler/test/verifytrim.jl:
```diff
@@ -18,31 +18,42 @@ let infos = Any[]
     desc = desc.second
     @test desc isa CallMissing
-    @test occursin("finalizer", desc.desc)
+    @test occursin("cfunction", desc.desc)
     repr = sprint(verify_print_error, desc, parents)
     @test occursin(
-        r"""^unresolved finalizer registered from statement \(Core.finalizer\)\(Base.final_shred!, %new\(\)::Base.SecretBuffer\)::Nothing
```
We should put back this test, for code coverage sake, both for SecretBuffer (working) and for something else (brokeN). If necessary (since this is presumably fixed), we can change the initial inferred call to something like `Core.svec(finalizer, Tuple{identity, Any})`

topolarity replied:
The working test for SecretBuffer is on https://github.com/JuliaLang/julia/pull/58014/files/0578395d9c980f496b188a383cc9266bce86d689#diff-e3e801c6894fd826c1162f7037c483ed0df8955c10a29b25cbaa22d538482e7bR53

but adding coverage for a failed `finalizer` verification sounds like an excellent idea

vtjnash commented on Compiler/src/verifytrim.jl:
```diff
@@ -219,9 +221,17 @@ function verify_codeinstance!(codeinst::CodeInstance, codeinfo::CodeInfo, inspec
-                        warn = true
                     end
                 else
                     error = "unresolved call to builtin"
```
Could you move this to the top of the `<: Builtin` code? Looks like we have a few places that would be left with the default message instead ("unresolved call"), which may be mildly annoying

Could you also add a check `if !isconcretetype(ftyp); error = "unknown Builtin"; end`? It is unlikely someone manages to infer as just Builtin or a Union, but upon further reflection, it seems like it would probably be confusing that `ftyp isa` would trigger on a random case statement and end up trying to explain why this unknown function would not make be a valid call to specifically `_apply_iterate`
